vault/identity_store_structs.go

```diff
@@ -0,0 +1,184 @@
+	storagePacker *storagePacker
+}
+
+// entityStorageEntry represents an entity that gets persisted and indexed.
+// Entity is fundamentally composed of zero or many identities.
+type entityStorageEntry struct {
+	// Identities are the identities that this entity is made of. This can be
+	// empty as well to favor being able to create the entity first and then
+	// incrementally adding identities.
+	Identities []*identityIndexEntry `json:"identities" structs:"identities" mapstructure:"identities"`
+
+	// ID is the unique identifier of the entity which always be a UUID. This
+	// should never be allowed to be updated.
+	ID string `json:"id" structs:"id" mapstructure:"id"`
+
+	// Name is a unique identifier of the entity which is intended to be
+	// human-friendly. The default name might not be human friendly since it
+	// gets suffixed by a UUID, but it can optionally be updated, unlike the ID
+	// field.
+	Name string `json:"name" structs:"name" mapstructure:"name"`
+
+	// Metadata represents the explicit metadata which is set by the clients.
+	// This is useful to tie any information pertaining to the identity. This
+	// is a non-unique field of entity, meaning multiple entities can have the
+	// same metadata set. Entities will be indexed based on this explicit
+	// metadata. This enables virtual groupings of entities based on its
+	// metadata.
+	Metadata map[string]string `json:"metadata" structs:"metadata" mapstructure:"metadata"`
+
+	// CreationTime is the time at which this entity is first created.
+	CreationTime time.Time `json:"creation_time" structs:"creation_time" mapstructure:"creation_time"`
+
+	// LastUpdateTime is the most recent time at which the properties of this
+	// entity got modified. This is helpful in filtering out entities based on
+	// its age and to take action on them, if desired.
+	LastUpdateTime time.Time `json:"last_update_time" structs:"last_update_time" mapstructure:"last_update_time"`
+
+	// MergedEntities are the entities which got merged to this one. Entities
+	// will be indexed based on all the entities that got merged into it. This
+	// helps to apply the actions on this entity on the tokens that are merged
+	// to the merged entities. Merged entities will be deleted entirely and
+	// this is the only trackable trail of its earlier presence.
+	MergedEntities []string `json:"merged_entities" structs:"merged_entities" mapstructure:"merged_entities"`
+
+	// Implicit helps to distinguish between entities that Vault created
+	// automatically vs entities that were registered explicitly.
+	Implicit bool `json:"implicit" structs:"implicit" mapstructure:"implicit"`
+
+	// ImplicitMetadata is the set of information Vault internally assigns to
+	// the entities. This field might serve as a basis for implicit groups of
+	// entities which is planned for the coming releases.
+	ImplicitMetadata map[string]string `json:"implicit_metadata" structs:"implicit_metadata" mapstructure:"implicit_metadata"`
```

Comment:
This value seems like it's putting the cart before the horse. If there's no actual use case for this now I'd prefer to leave it off, especially since it might be better handled via a `map[string]interface{}` or a custom `struct`